pkg/image/registry.go

```diff
@@ -42,49 +43,102 @@ func AddDefaultRegistry(imageToChange string, registry string) (string, error) {
 	return toAdd, nil
 }
 
-// ParseOstreeReference returns two strings.  First, the ostree
-// reference without a tag.  Second, a reference that is usable
-// by Podman compatible container runtimes.  For example,
+// ParseOstreeReference returns three strings: the ostree transport, registry,
+// and tag for an ostree container image.  For example,
 // "ostree-unverified-image:container-registry.oracle.com/olcne/ock-ostree:1.30"
-// returns "ostree-unverified-image:container-registry.oracle.com/olcne/ock-ostree"
-// and container-registry.oracle.com/olcne/ock-ostree:1.30
-func ParseOstreeReference(img string) (string, string, error) {
+// returns "ostree-unverified-image", "container-registry.oracle.com/olcne/ock-ostree"
+// and "1.30".  It is not necessary that a tag is present, but all the rest of the
+// fields are required.
+func ParseOstreeReference(img string) (string, string, string, error) {
 	// Important stuff is colon delimited.
 	fields := strings.Split(img, ":")
 
-	// At the very least there needs to be a registry
-	// and a tag.  More, actually, but that is all checked
-	// later on.
+	// At the very least there needs to be a transport and a registry.
 	if len(fields) < 2 {
-		return "", "", fmt.Errorf("%s is not a valid ostree image reference", img)
+		return "", "", "", fmt.Errorf("%s is not a valid ostree image reference", img)
 	}
 
-	switch fields[0] {
+	ostreeTransport := fields[0]
+	switch ostreeTransport {
 	case "ostree-unverified-image", "ostree-image-signed":
 		fields = fields[1:]
+		ostreeTransport = fmt.Sprintf("%s:%s", ostreeTransport, fields[0])
 		switch fields[0] {
 		case "registry":
 			fields = fields[1:]
 		case "docker":
 			// strip off the "//"
+			ostreeTransport = fmt.Sprintf("%s://", ostreeTransport)
+			fields = fields[1:]
 			fields[0] = fields[0][2:]
 		default:
-			return "", "", fmt.Errorf("%s is not a valid ostree image reference", img)
+			return "", "", "", fmt.Errorf("%s is not a valid ostree image reference", img)
 		}
 	case "ostree-unverified-registry":
+		ostreeTransport = fields[0]
 		fields = fields[1:]
 	case "ostree-remote-image":
-		if len(fields) < 3 {
-			return "", "", fmt.Errorf("%s is not a valid ostree image reference", img)
+		if len(fields) < 4 {
+			return "", "", "", fmt.Errorf("%s is not a valid ostree image reference", img)
+		}
+
+		ostreeTransport = fmt.Sprintf("%s:%s", ostreeTransport, fields[1])
+		switch fields[2] {
+		case "registry":
+			ostreeTransport = fmt.Sprintf("%s:%s", ostreeTransport, fields[2])
+		case "docker":
+			ostreeTransport = fmt.Sprintf("%s:docker://", ostreeTransport)
+			if !strings.HasPrefix(fields[3], "//") {
+				return "", "", "", fmt.Errorf("%s is not a valid ostree image reference", img)
+			}
+			fields[3] = fields[3][2:]
 		}
-		fields = fields[2:]
+		fields = fields[3:]
 	default:
```

Comment:
The steps used to split the ostree image in the default case are very similar to the steps in review here:

https://github.com/oracle-cne/ocne/blob/57522142f7fc364b5371f4ab4e6a2841595d8ab2/pkg/image/image.go#L398

The function in the PR has unit testing and offers specific error messages. 

Should we change either of these code blocks to match each other?

Reply:
It would be great if this function used `SplitImage` instead of rolling its own OCI image splitter.  We could also change the return value of this function to an ImageInfo.